### PR TITLE
Add support for \x?? and improve escapes

### LINKED ISSUE #82: Serialization of Control Characters
Currently ASCII control characters like  `'\x07'` are directly serialized. I expect them to be escaped like in Rust Code.

### PR DESCRIPTION
**Pull request by torkleyy**
~~**This PR seems to contain a regression.**~~

Fixes #82

### CODE REVIEW

File: tests/escape.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    let s: String = to_string(&t).unwrap();
+
+    println!("Serialized: \n\n{}\n\n", s);
+
+    assert_eq!(from_str(&s), Ok(t));
+}
+
+#[test]
+fn test_ascii_10() {
+    check_same("\u{10}".to_owned());
```
Comment (torkleyy) [inline tests/escape.rs]: That's what causes the issue: the parser reads this as `\x01` instead of `\x10`.

**Comment by torkleyy** [comment]
This was a stupid mistake, fixed it now :)

This is ready for review.

**Comment by pyfisch** [comment]
Thanks. How about using [`char::escape_debug`](https://doc.rust-lang.org/std/primitive.char.html?search=#method.escape_debug) instead for both chars and strings.
There is also a [`u32::from_str_radix`](https://doc.rust-lang.org/std/primitive.u32.html#method.from_str_radix) to parse hexadecimal numbers.
Rust itself only [support very few special escapes](https://doc.rust-lang.org/reference/tokens.html#character-escapes) and RON should IMHO not invent new ones (or copy them from JSON).

**Comment by torkleyy** [comment]
That's a good idea, I'll try it out tomorrow.

**Comment by pyfisch** [comment]
@torkleyy That I like to be [able to read Unicode characters](https://play.rust-lang.org/?gist=09c795f00833601793dd21c892c44c7c&version=stable).

**Comment by torkleyy** [comment]
Ah, thanks for the explanation :)

**Comment by kvark** [comment]
bors delegate=pyfisch

**Comment by torkleyy** [comment]
I finally finished this :) r? @pyfisch @kvark

File: src/parse.rs [hunk 2]
```diff
@@ -421,82 +429,93 @@ impl<'a> Bytes<'a> {
...
+            n = n << 4;
+            let byte = self.eat_byte()?;
+            let decoded = self.decode_hex(byte)?;
+            n |= decoded;
+        }
+
+        Ok(n)
+    }
+
+    /*
```
Comment (pyfisch) [inline src/parse.rs]: Remove the commented out code.

File: src/ser/mod.rs [hunk 3]
```diff
@@ -156,6 +156,12 @@ impl Serializer {
                 .extend((0..pretty.indent).map(|_| config.indentor.as_str()));
         }
     }
+
+    fn serialize_escaped_str(&mut self, value: &str) {
+        self.output += "\"";
+        self.output += &value.chars().flat_map(|c| c.escape_default()).collect::<String>();
```
Comment (pyfisch) [inline src/ser/mod.rs]: I would avoid the string allocation and immediately push the characters to the output.

File: tests/escape.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+fn test_ascii_chars() {
+    (1..128).into_iter().flat_map(from_u32).for_each(check_same)
+}
+
+#[test]
+fn test_ascii_string() {
+    let s: String = (1..128).into_iter().flat_map(from_u32).collect();
+
+    check_same(s);
+}
```
Comment (pyfisch) [inline tests/escape.rs]: Some test with non-ascii characters would be nice.

For example `♠` and `ß`. (They should not be escaped)

File: src/parse.rs [hunk 5]
```diff
@@ -421,82 +429,93 @@ impl<'a> Bytes<'a> {
...
+                    bytes = bytes << 4;
+                    bytes |= byte as u32;
+                }
 
-                        if n2 < 0xDC00 || n2 > 0xDFFF {
-                            return self.err(ParseError::InvalidEscape);
-                        }
+                if bytes == 0 {
+                    return self.err(ParseError::InvalidEscape(
+                        "Null character (\\u{0}) not allowed",
```
Comment (pyfisch) [inline src/parse.rs]: Why is the null character not allowed? "foo\0bar" is a valid string in Rust.

File: src/parse.rs [hunk 6]
```diff
@@ -421,82 +429,93 @@ impl<'a> Bytes<'a> {
...
+    }
 
     fn parse_str_escape(&mut self, store: &mut Vec<u8>) -> Result<()> {
         use std::iter::repeat;
 
         match self.eat_byte()? {
+            b'\'' => store.push(b'\''),
             b'"' => store.push(b'"'),
             b'\\' => store.push(b'\\'),
             b'b' => store.push(b'\x08'),
```
Comment (pyfisch) [inline src/parse.rs]: Some of these escapes are no standard Rust escapes.

File: src/ser/mod.rs (hunk 3 above)
Comment (torkleyy) [reply src/ser/mod.rs]: Err right, I was trying it out and forgot about it then. Thanks!

File: src/parse.rs (hunk 5 above)
Comment (torkleyy) [reply src/parse.rs]: I don't think we should allow it in RON though, since that would make usage in e.g. C really difficult.

**Comment by torkleyy** [comment]
Okay, addressed all comments except one (about `\0`).

**Comment by pyfisch** [comment]
I don't like the current handling of NUL. Currently the string "foo\0bar" will successfully be serialized. But it will fail deserialization.

My assumption is that I can serialize any Rust data type with RON and deserialize it again and get an equivalent representation. Special casing strings with a NUL character breaks this assumption.

**Comment by torkleyy** [comment]
You're right that serialization and deserialization should allow the same things. If we decide to exclude `\0` I will add a check to the serialization.

@kvark @ozkriff @Xaeroxe What do you think about this?

My main concern is that adding support for `\0` would make using RON in e.g. C really difficult. Some parsers for e.g. JSON have to use special hacks and callbacks to make `\0` inside a string work.

**Comment by pyfisch** [comment]
My primary concern is that if we exclude `\0` somewhere in Servo a string with a `\0` will occur and on serialization RON panics. In this case servo will probably be forced to use a different format  without this quirk. (I *think* the current use cases for debugging in Servo and WebRender do not allow strings with `\0` but it is better to be safe)

**Comment by torkleyy** [comment]
It doesn't panic, just gives you an error. And really, it doesn't make _any_ sense to include `\0` in a string. If a format supports that it's a bug, not a feature.

**Comment by kvark** [comment]
C interop and serialization are somewhat orthogonal. I could see a problem with other languages implementing RON support and treating this differently. For that we should have NULL handling defined as implementation dependent. There are cases in OpenGL/Vulkan spec where they do the same: it's not undefined behavior, and it's not standard, it's a defined per implementation behavior.

TL;DR: we should support `\0` in *this* implementation.

**Comment by pyfisch** [comment]
bors delegate+

Merge it after squashing.

**Comment by torkleyy** [comment]
bors r=pyfisch
